x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/allocation/TrainedModelAllocation.java

```diff
@@ -0,0 +1,202 @@
+        nodeRoutingTable = in.readOrderedMap(StreamInput::readString, RoutingStateAndReason::new);
+    }
+
+    public boolean routedToNode(String nodeId) {
+        return nodeRoutingTable.containsKey(nodeId);
+    }
+
+    public Map<String, RoutingStateAndReason> getNodeRoutingTable() {
+        return Collections.unmodifiableMap(nodeRoutingTable);
+    }
+
+    public StartTrainedModelDeploymentAction.TaskParams getTaskParams() {
+        return taskParams;
+    }
+
+    // TODO add support for other roles?
+    // NOTE, whatever determines allocation should not be dynamically setable on the node
+    // Otherwise allocation logic might fail
+    public boolean canAllocateToNode(DiscoveryNode node) {
+        return node.getRoles().contains(DiscoveryNodeRole.ML_ROLE);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        TrainedModelAllocation that = (TrainedModelAllocation) o;
+        return Objects.equals(nodeRoutingTable, that.nodeRoutingTable) && Objects.equals(taskParams, that.taskParams);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(nodeRoutingTable, taskParams);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(TASK_PARAMETERS.getPreferredName(), taskParams);
+        builder.field(ROUTING_TABLE.getPreferredName(), nodeRoutingTable);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        taskParams.writeTo(out);
+        out.writeMap(nodeRoutingTable, StreamOutput::writeString, (o, w) -> w.writeTo(o));
+    }
+
+    public static class Builder {
+        private final Map<String, RoutingStateAndReason> nodeRoutingTable;
+        private final StartTrainedModelDeploymentAction.TaskParams taskParams;
+        private boolean isChanged;
+
+        public static Builder fromAllocation(TrainedModelAllocation allocation) {
+            return new Builder(allocation.taskParams, allocation.nodeRoutingTable);
+        }
+
+        public static Builder empty(StartTrainedModelDeploymentAction.TaskParams taskParams) {
+            return new Builder(taskParams);
+        }
+
+        private Builder(StartTrainedModelDeploymentAction.TaskParams taskParams, Map<String, RoutingStateAndReason> nodeRoutingTable) {
+            this.taskParams = taskParams;
+            this.nodeRoutingTable = new LinkedHashMap<>(nodeRoutingTable);
+        }
+
+        private Builder(StartTrainedModelDeploymentAction.TaskParams taskParams) {
+            this.nodeRoutingTable = new LinkedHashMap<>();
+            this.taskParams = taskParams;
+        }
+
+        public Builder addNewRoutingEntry(String nodeId) {
```

Comment:
@dimitris-athanasiou My idea was to restrict how a NEW entry could be added. Either it is `Initializing` or `Failing`, we shouldn't allow a `STOPPING` NEW routing entry

Reply:
@davidkyle this is for the `master` node. Its the only thing that can create routing entries. But, since it is also the thing that checks if there is memory available on the node, it needs to indicate if it failed to allocate due to memory restrictions.

Adding a failed route does this.